ggml/src/ggml-alloc.c

```diff
@@ -316,10 +360,93 @@ static void ggml_dyn_tallocr_free(struct ggml_dyn_tallocr * alloc) {
 }
 
 static size_t ggml_dyn_tallocr_max_size(struct ggml_dyn_tallocr * alloc) {
-    return alloc->max_size;
+    return alloc->max_size[alloc->n_chunks - 1];
+}
+
+
+// virtual buffer with contiguous memory range, split into multiple backend buffers (chunks)
+
+struct vbuffer {
+    ggml_backend_buffer_type_t buft;
+    ggml_backend_buffer_t chunks[GGML_VBUFFER_MAX_CHUNKS];
+};
+
+static struct vbuffer * ggml_vbuffer_new(ggml_backend_buffer_type_t buft) {
+    struct vbuffer * buf = calloc(1, sizeof(struct vbuffer));
+    buf->buft = buft;
+    memset(buf->chunks, 0, sizeof(buf->chunks));
+    return buf;
+}
+
+static void ggml_vbuffer_free_chunks(struct vbuffer * buf) {
+    for (int i = 0; i < GGML_VBUFFER_MAX_CHUNKS; ++i) {
+        ggml_backend_buffer_free(buf->chunks[i]);
+        buf->chunks[i] = NULL;
+    }
+}
+
+static void ggml_vbuffer_free(struct vbuffer * buf) {
+    if (buf == NULL) {
+        return;
+    }
+    ggml_vbuffer_free_chunks(buf);
+    free(buf);
+}
+
+static int ggml_vbuffer_n_chunks(struct vbuffer * buf) {
+    int n = 0;
+    while (n < GGML_VBUFFER_MAX_CHUNKS && buf->chunks[n]) n++;
+    return n;
+}
+
+static size_t ggml_vbuffer_size(struct vbuffer * buf) {
+    size_t size = 0;
+    for (int i = 0; i < GGML_VBUFFER_MAX_CHUNKS && buf->chunks[i]; ++i) {
+        size += ggml_backend_buffer_get_size(buf->chunks[i]);
+    }
+    return size;
+}
+
+static bool ggml_vbuffer_alloc(struct vbuffer * buf, const struct ggml_dyn_tallocr * talloc, enum ggml_backend_buffer_usage usage) {
+    for (int n = 0; n < talloc->n_chunks; n++) {
+        size_t chunk_size = talloc->max_size[n];
+        if (n > 0) {
+            chunk_size -= talloc->max_size[n - 1];
+        }
+        buf->chunks[n] = ggml_backend_buft_alloc_buffer(buf->buft, chunk_size);
+        if (buf->chunks[n] == NULL) {
+            ggml_vbuffer_free_chunks(buf);
+            return false;
+        }
+        ggml_backend_buffer_set_usage(buf->chunks[n], usage);
+    }
+    return true;
+}
+
+static void ggml_vbuffer_tensor_alloc(struct vbuffer * buf, struct ggml_tensor * tensor, size_t offset) {
+    size_t chunk_index = 0, chunk_offset = offset;
+    while (chunk_index < GGML_VBUFFER_MAX_CHUNKS && buf->chunks[chunk_index]) {
+        size_t chunk_size = ggml_backend_buffer_get_size(buf->chunks[chunk_index]);
+        if (chunk_offset < chunk_size) {
+            break;
+        }
+        chunk_offset -= chunk_size;
+        chunk_index++;
+    }
```

Comment:
I don't understand why this loop is necessary. Why not store the chunk index in `tensor_alloc`?

Reply:
Conceptually, I kept any notion of chunks out of gallocr, it continues to operate with global offset only. This requires reconstructing it into chunk index + local offset here. I thought it would be easier to reason about gallocr that way, but maybe was overthinking things.

> Why not store the chunk index in `tensor_alloc`?

It needs to pass through `hash_node` too, and is copied in various places. In that case I would replace all `size_t offset` with a pair (chunk_index, local_offset), which can then be treated in much the same way, and won't require these reconstruction loops.

That sounds good to you?